DEV: Remove gifsicle dependency

nbianca opened the pull request.

xfalcox commented:
We should add a PSA to the announcements when we merge this.

nbianca commented:
This is ready now. FastImage v2.2.0 contains my additions to the gem.

ZogStriP commented on app/jobs/onceoff/rebake_gif_images.rb:
```diff
@@ -0,0 +1,43 @@
+    def execute_onceoff(args)
+      Upload.where("original_filename LIKE '%.gif'").find_each do |upload|
+        uri = Discourse.store.path_for(upload) || upload.url
+        upload.update!(animated: FastImage.animated?(uri))
```
Should we catch any exceptions here?

nbianca replied:
I do not think so. What exceptions do you have in mind?

FastImage should not throw (there is an option for that, disabled by default) and if AR throws then there is something more serious going on and we should see that in logs. If a onceoff job throws, it seems that it is repeated until it succeeds, which  is desired behavior.

ZogStriP commented on app/jobs/onceoff/rebake_gif_images.rb:
```diff
@@ -0,0 +1,43 @@
+          if path = Discourse.store.path_for(user.uploaded_avatar)
+            file = File.new(path)
+          else
+            find = Discourse.store.download(user.uploaded_avatar)
```
I think you meant `file` here?

```suggestion
            file = Discourse.store.download(user.uploaded_avatar)
```

nbianca replied:
Yes, you are right. 🤦🏻‍♀️

ZogStriP commented on app/jobs/onceoff/rebake_gif_images.rb:
```diff
@@ -0,0 +1,43 @@
+          user.uploaded_avatar.update!(sha1: SecureRandom.hex(20))
+
+          upload = UploadCreator.new(file, user.uploaded_avatar.original_filename, type: "avatar").create_for(user.id)
+          user.update!(uploaded_avatar: upload)
```
Should we catch any exceptions here?

nbianca replied:
Same as above.

ZogStriP reviewed (approved):
LGTM 👍 

Should definitely come with a PSA on meta as well 😉

nbianca commented:
I simplified the logic for rebaking animated GIF images and moved it to a scheduled job (it was a once-off job before). The reason for this change is to split the load of downloading and parsing every GIF images over a longer time period. Also, it no longer recreates uploads as a simple recreation of the optimized images is enough (these are generally used).

SamSaffron commented on app/jobs/scheduled/rebake_gif_images.rb:
```diff
@@ -0,0 +1,23 @@
+
+    def execute(args)
+      Upload
+        .where("original_filename LIKE '%.gif'")
```
Should we use `.where("extension = 'gif' || original_filename like '%.gif' ")` ... gifs can sneak in with the wrong filename we rely on actual content I think.

SamSaffron replied:
It can be null though in some cases

nbianca replied:
That sounds like a good idea. :+1:

SamSaffron commented:
hmmm if we are not using "animated" anywhere why bother backfilling now? I think once we actually use this column it makes sense to have a back filler ... but, is there any point doing this early?

nbianca commented on lib/tasks/uploads.rake:
```diff
@@ -1194,3 +1194,23 @@ task "uploads:fix_missing_s3" => :environment do
+  RailsMultisite::ConnectionManagement.each_connection do |db|
+    puts "Backfilling #{db}..."
+    Upload
+      .where('original_filename LIKE \'%.gif\'')
```
I removed the backfilling Rake task completely. We do not need one now to force the process and we already have the scheduled job anyway...

ZogStriP commented on app/jobs/scheduled/update_animated_uploads.rb:
```diff
@@ -0,0 +1,23 @@
+        .where("extension = 'gif' OR (extension IS NULL AND original_filename LIKE '%.gif')")
+        .where(animated: nil)
+        .limit(MAX_PROCESSED_GIF_IMAGES)
+        .find do |upload|
```
Shouldn't that be `each` instead of `find`?
